--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/QtUI/clients_table.py.orig	2019-09-27 12:41:53.036513316 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/QtUI/clients_table.py	2019-09-27 12:42:18.637022014 -0400
@@ -20,9 +20,9 @@
 #-*- coding: utf-8 -*-
 
 import sys, time
-from PyQt4.QtCore import *
-from PyQt4 import QtDBus, QtGui
-from PyQt4.QtGui import QColor
+from PyQt5.QtCore import *
+from PyQt5 import QtDBus, QtGui
+from PyQt5.QtGui import QColor
 
 class ClientTableModel(QAbstractTableModel):
     '''
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/QtUI/exports_table.py.orig	2019-09-27 12:41:59.275637290 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/QtUI/exports_table.py	2019-09-27 12:42:22.021089257 -0400
@@ -19,9 +19,9 @@
 # Author: Jim Lieb <jlieb@panasas.com>
 #-*- coding: utf-8 -*-
 import sys, time
-from PyQt4.QtCore import *
-from PyQt4 import QtDBus, QtGui
-from PyQt4.QtGui import QColor
+from PyQt5.QtCore import *
+from PyQt5 import QtDBus, QtGui
+from PyQt5.QtGui import QColor
 
 class ExportTableModel(QAbstractTableModel):
     '''
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/QtUI/log_settings.py.orig	2019-09-27 12:42:03.307717408 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/QtUI/log_settings.py	2019-09-27 12:42:26.062169555 -0400
@@ -19,8 +19,8 @@
 # Author: Jim Lieb <jlieb@panasas.com>
 #-*- coding: utf-8 -*-
 from __future__ import print_function
-from PyQt4.QtCore import *
-from PyQt4 import QtGui
+from PyQt5.QtCore import *
+from PyQt5 import QtGui
 from Ganesha.QtUI.ui_log_dialog import Ui_LogSettings
 
 class DebugLevelDelegate(QtGui.QItemDelegate):
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/admin.py.orig	2019-09-27 12:41:03.790534769 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/admin.py	2019-09-27 12:41:28.637028484 -0400
@@ -19,7 +19,7 @@
 # Author: Jim Lieb <jlieb@panasas.com>
 #-*- coding: utf-8 -*-
 
-from PyQt4 import QtCore, QtDBus
+from PyQt5 import QtCore, QtDBus
 
 class AdminInterface(QtDBus.QDBusAbstractInterface):
     '''
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/export_mgr.py.orig	2019-09-27 12:41:09.574649703 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/export_mgr.py	2019-09-27 12:41:32.820111603 -0400
@@ -20,8 +20,8 @@
 #-*- coding: utf-8 -*-
 
 import sys, time
-from PyQt4.QtCore import *
-from PyQt4 import QtDBus, QtGui
+from PyQt5.QtCore import *
+from PyQt5 import QtDBus, QtGui
 from collections import namedtuple
 
 Export = namedtuple('Export',
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/log_mgr.py.orig	2019-09-27 12:41:15.542768292 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/log_mgr.py	2019-09-27 12:41:36.229179345 -0400
@@ -19,7 +19,7 @@
 # Author: Jim Lieb <jlieb@panasas.com>
 #-*- coding: utf-8 -*-
 
-from PyQt4 import QtCore, QtDBus
+from PyQt5 import QtCore, QtDBus
 
 ADMIN_OBJECT = '/org/ganesha/nfsd/admin'
 PROP_INTERFACE = 'org.freedesktop.DBus.Properties'
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/client_mgr.py.orig	2019-09-27 12:43:03.522913305 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/Ganesha/client_mgr.py	2019-09-27 12:43:12.983100960 -0400
@@ -19,7 +19,7 @@
 # Author: Jim Lieb <jlieb@panasas.com>
 #-*- coding: utf-8 -*-
 
-from PyQt4 import QtCore, QtDBus
+from PyQt5 import QtCore, QtDBus
 from collections import namedtuple
 
 Client = namedtuple('Client',
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/ganesha-admin.py.orig	2019-09-27 12:38:51.304902206 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/ganesha-admin.py	2019-09-27 12:39:57.279213153 -0400
@@ -21,8 +21,8 @@
 #-*- coding: utf-8 -*-
 from __future__ import print_function
 import sys
-from PyQt4 import QtCore, QtDBus
-from PyQt4.QtGui import QApplication
+from PyQt5 import QtCore, QtDBus
+from PyQt5.QtGui import QApplication
 from dbus.mainloop.qt import DBusQtMainLoop
 from Ganesha.admin import AdminInterface
 
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/ganeshactl.py.orig	2019-09-27 12:38:59.017055448 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/ganeshactl.py	2019-09-27 12:40:18.620637219 -0400
@@ -1,6 +1,6 @@
 #!/usr/bin/python2
 #
-# ganeshactl.py - PyQt4 GUI tool for admin of nfs-ganesha.
+# ganeshactl.py - PyQt5 GUI tool for admin of nfs-ganesha.
 #
 # Copyright (C) 2014 Panasas Inc.
 #
@@ -26,7 +26,7 @@
 
 import sys
 
-from PyQt4 import QtCore, QtGui, QtDBus 
+from PyQt5 import QtCore, QtGui, QtDBus 
 from Ganesha.QtUI.ui_main_window import Ui_MainWindow
 from Ganesha.admin import AdminInterface
 from Ganesha.export_mgr import ExportMgr
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/manage_clients.py.orig	2019-07-22 15:17:01.000000000 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/manage_clients.py	2019-09-27 12:40:33.170926341 -0400
@@ -22,8 +22,8 @@
 from __future__ import print_function
 import sys
 import time
-from PyQt4 import QtCore, QtDBus
-from PyQt4.QtGui import QApplication
+from PyQt5 import QtCore, QtDBus
+from PyQt5.QtGui import QApplication
 from dbus.mainloop.qt import DBusQtMainLoop
 from Ganesha.client_mgr import ClientMgr
 
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/manage_exports.py.orig	2019-09-27 12:39:29.560662368 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/manage_exports.py	2019-09-27 12:40:38.596034141 -0400
@@ -22,8 +22,8 @@
 from __future__ import print_function
 import sys
 import time
-from PyQt4 import QtCore, QtDBus
-from PyQt4.QtGui import QApplication
+from PyQt5 import QtCore, QtDBus
+from PyQt5.QtGui import QApplication
 from dbus.mainloop.qt import DBusQtMainLoop
 from Ganesha.export_mgr import ExportMgr
 
--- nfs-ganesha-2.8.2/src/scripts/ganeshactl/manage_logger.py.orig	2019-09-27 12:39:36.692804090 -0400
+++ nfs-ganesha-2.8.2/src/scripts/ganeshactl/manage_logger.py	2019-09-27 12:40:46.004181345 -0400
@@ -22,8 +22,8 @@
 from __future__ import print_function
 import sys
 import time
-from PyQt4 import QtCore, QtDBus
-from PyQt4.QtGui import QApplication
+from PyQt5 import QtCore, QtDBus
+from PyQt5.QtGui import QApplication
 from dbus.mainloop.qt import DBusQtMainLoop
 from Ganesha.log_mgr import LogManager
 
