--- nfs-ganesha-2.4.0/src/MainNFSD/nfs_main.c.orig	2016-09-26 08:20:46.180000000 -0400
+++ nfs-ganesha-2.4.0/src/MainNFSD/nfs_main.c	2016-09-26 09:03:09.536000000 -0400
@@ -167,8 +167,8 @@
 		case 'v':
 		case '@':
 			/* A litlle backdoor to keep track of binary versions */
-			printf("%s compiled on %s at %s\n", exec_name, __DATE__,
-			       __TIME__);
+			/* printf("%s compiled on %s at %s\n", exec_name, __DATE__,
+			       __TIME__); */
 			printf("Release = V%s\n", GANESHA_VERSION);
 			printf("Release comment = %s\n", VERSION_COMMENT);
 			printf("Git HEAD = %s\n", _GIT_HEAD_COMMIT);
@@ -246,7 +246,7 @@
 		 "%s Starting: %s",
 		 exec_name,
 		 "Ganesha Version " _GIT_DESCRIBE ", built at "
-		 __DATE__ " " __TIME__ " on " BUILD_HOST);
+		 /* __DATE__ " " __TIME__ */ " on " BUILD_HOST);
 
 	nfs_check_malloc();
 
--- nfs-ganesha-2.4.0/src/MainNFSD/nfs_lib.c.orig	2016-09-26 08:34:11.884000000 -0400
+++ nfs-ganesha-2.4.0/src/MainNFSD/nfs_lib.c	2016-09-26 08:49:22.764000000 -0400
@@ -128,7 +128,7 @@
 		 "%s Starting: %s",
 		 exec_name,
 		 "Ganesha Version " _GIT_DESCRIBE ", built at "
-		 __DATE__ " " __TIME__ " on " BUILD_HOST);
+		 /* __DATE__ " " __TIME__ */ " on " BUILD_HOST);
 
 	nfs_check_malloc();
 
