--- nfs-ganesha-2.5.0/src/scripts/systemd/nfs-ganesha.service.orig	2017-06-27 15:28:51.992000000 -0400
+++ nfs-ganesha-2.5.0/src/scripts/systemd/nfs-ganesha.service	2017-06-27 15:29:13.864000000 -0400
@@ -23,7 +23,7 @@
 [Service]
 Type=forking
 Environment="NOFILE=1048576"
-EnvironmentFile=-/run/sysconfig/ganesha
+EnvironmentFile=-/etc/default/nfs-ganesha
 ExecStart=/bin/bash -c "${NUMACTL} ${NUMAOPTS} /usr/bin/ganesha.nfsd ${OPTIONS} ${EPOCH}"
 ExecStartPost=-/bin/bash -c "prlimit --pid $MAINPID --nofile=$NOFILE:$NOFILE"
 ExecReload=/bin/dbus-send --system   --dest=org.ganesha.nfsd --type=method_call /org/ganesha/nfsd/admin  org.ganesha.nfsd.admin.reload
--- nfs-ganesha-2.5.0/src/scripts/nfs-ganesha-config.sh.orig	2017-07-13 16:26:14.788000000 -0400
+++ nfs-ganesha-2.5.0/src/scripts/nfs-ganesha-config.sh	2017-07-13 19:23:50.616000000 -0400
@@ -7,11 +7,17 @@
 #
 
 CONFIGFILE=/etc/sysconfig/ganesha
-if test -r ${CONFIGFILE}; then
+RUNCONFIG=/run/sysconfig/ganesha
+if [ ! -e $(command dirname ${CONFIGFILE} 2>/dev/null) ]; then
+	CONFIGFILE=/etc/ganesha/ganesha
+	RUNCONFIG=/etc/default/ganesha
+fi
+
+if [ -r ${CONFIGFILE} ]; then
 	. ${CONFIGFILE}
 	[ -x ${EPOCH_EXEC} ] &&  EPOCHVALUE=`${EPOCH_EXEC}`
 
-	mkdir -p /run/sysconfig
+	mkdir -p $(command dirname ${RUNCONFIG} 2>/dev/null)
 	{
 		cat ${CONFIGFILE}
 		[ -n "${EPOCHVALUE}" ] && echo EPOCH=\"-E $EPOCHVALUE\"
@@ -19,8 +25,8 @@
 		# Set NUMA options if numactl is present
 		NUMACTL=$(command -v numactl 2>/dev/null)
 		if [ -n "${NUMACTL}" ]; then
-			echo NUMACTL=$NUMACTL
+			echo NUMACTL=${NUMACTL}
 			echo NUMAOPTS=--interleave=all
 		fi
-	} > /run/sysconfig/ganesha
+	} > ${RUNCONFIG}
 fi
